k9mail-library/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.java

```diff
@@ -566,17 +572,38 @@ private void sendMessageTo(List<String> addresses, Message message)
         Address[] from = message.getFrom();
         try {
             String fromAddress = from[0].getAddress();
-            if (m8bitEncodingAllowed) {
-                executeCommand("MAIL FROM:<%s> BODY=8BITMIME", fromAddress);
+            if (pipeliningSupported) {
+                pipelinedCommand = new LinkedList<>();
+                if (m8bitEncodingAllowed) {
+                    executeCommandPipelined("MAIL FROM:<%s> BODY=8BITMIME", fromAddress);
+                } else {
+                    executeCommandPipelined("MAIL FROM:<%s>", fromAddress);
+                }
+
+                for (String address : addresses) {
+                    executeCommandPipelined("RCPT TO:<%s>", address);
+                }
+
+                executeCommandPipelined("DATA");
+                CommandResponse commandResponse = readPipelinedResponse();
+                if (commandResponse.replyCode != 354) {
+                    String replyText = TextUtils.join(" ", commandResponse.results);
+                    throw new NegativeSmtpReplyException(commandResponse.replyCode, replyText);
```

Comment:
this seems odd, and also breaks enhanced status codes. Maybe make readPipelinedResponse throw if the last command is a negative reply? that aligns well with its return value, which is also only the last response.